osquery/sql/virtual_table.cpp

```diff
@@ -582,59 +906,78 @@ static int xFilter(sqlite3_vtab_cursor* pVtabCursor,
   pCur->n = pCur->data.size();
   return SQLITE_OK;
 }
+
+struct sqlite3_module* getVirtualTableModule(const std::string& table_name,
+                                             bool extension) {
+  static std::unordered_map<std::string, struct sqlite3_module> modules;
```

Comment:
This is done to actually avoid making the code messier (because we are reducing the scope as much as possible since it is not needed outside the dedicated interface function). This is also thread safe (as long as we compile with C++11 or better).

I can move this outside and remove the static if you prefer

Reply:
Makes sense :)

Reply:
Hi @alessandrogario I believe initialization 'static std::unordered_map<std::string, struct sqlite3_module> modules;' is thread safe but 'modules[table_name] = {};' is not thead safe
I.e. initialization of a function local static object is thread safe not the operations on them.

Reply:
Yes! I used the wrong terms; initialization is performed on a single thread (you can see the original code doesn't protect access to this variable)

Reply:
In the present code usage of variable 'module' is thread safe since -
1. It is initialized as static variable.
2. After initialization it is being used as read-only.
While the code you are proposing 'modules' variable is not read-only.

Reply:
You are right it is not accessed in read only, but I was under the assumption that initialization was single threaded, when in fact it is not. I'll address this issue. Thanks!

Reply:
"it does seem that everything is performed in a unique call from a single thread, so it doesn't matter whether you lock it or not."
I observed that the code is written with keeping thread safety in mind. Until now access to module object was being done in thread safe manner since it is being accessed as read-only. At present only single thread of control is accessing this code but it may not be the case in future.  

I believe attachlock will not be enough since it is synchronizing access to instance but modules object  can be accessed across the instances. Please correct me if I am wrong.

Reply:
You may be right! I've ended up reverting the last commit, and adding a mutex. Thanks!